libafl_frida/src/drcov_rt.rs

```diff
@@ -53,9 +53,13 @@ impl FridaRuntime for DrCovRuntime {
         let mut hasher = RandomState::with_seeds(0, 0, 0, 0).build_hasher();
         hasher.write(input.target_bytes().as_slice());
 
-        let filename = self
-            .coverage_directory
-            .join(format!("{:016x}.drcov", hasher.finish(),));
+        let hash = hasher.finish();
+        let mut filename = self.coverage_directory.join(format!("{hash:016x}.drcov"));
+        let mut i = 0;
+        while filename.exists() {
```

Comment:
This would still be race-y, right? We are doing something similar in the Corpus (but a better way would be... better)

Reply:
What's the goal here? Store all coverage entries, or keep a safe mapping of input -> coverage?

If it's just about keeping the coverage, why not hash the coverage instead of the input, and increase the hash size? The mapping of input->coverage could be symlinks, and it would have the advantage of keeping coverage files deduped.

Reply:
I'm absolutely agree with @domenukk and thank you @fabianfreyer.

I decided to do it a little differently. Firstly, empty coverages are now skipped, secondly, the file name consists of input data hashes and coverages. I made a compound name because I believe that an unforeseen situation may arise when the coverage may be different from the same input data (unstable harness).

Reply:
sounds good...

Reply:
I'd suggest on systems that support symlink semantics, to save the coverage to `{coverage_hash}.drcov` and make a symlink from `{input_hash}_{coverage_hash}.drcov` to it. That would have the following advantages:

- We're deduping coverage (if the hash is sparse enough)

Reply:
I tried to implement this. @fabianfreyer, you mean something like that?

<details>
  <summary>ls -l ./coverage/**</summary>
  
```
lrwxrwxrwx 1 root root   54 2023-10-04 10:07 ./coverage/014446d5e22ccab7_6449d035261ae158.drcov -> ./coverage/014446d5e22ccab7/6449d035261ae158.drcov
lrwxrwxrwx 1 root root   54 2023-10-04 10:07 ./coverage/02bd0d6773730397_990b7a01d108da6e.drcov -> ./coverage/02bd0d6773730397/990b7a01d108da6e.drcov
lrwxrwxrwx 1 root root   54 2023-10-04 10:07 ./coverage/067ccf3386f6dffd_fff39e5f01f8ebb1.drcov -> ./coverage/067ccf3386f6dffd/fff39e5f01f8ebb1.drcov
lrwxrwxrwx 1 root root   54 2023-10-04 10:07 ./coverage/132e025befe53cb9_8d7aca8322c310bf.drcov -> ./coverage/132e025befe53cb9/8d7aca8322c310bf.drcov
lrwxrwxrwx 1 root root   54 2023-10-04 10:07 ./coverage/1690bb2862eb85a0_635d5a3dfd59fcb1.drcov -> ./coverage/1690bb2862eb85a0/635d5a3dfd59fcb1.drcov
lrwxrwxrwx 1 root root   54 2023-10-04 10:07 ./coverage/17a47e74d0552750_be8552ced8f72971.drcov -> ./coverage/17a47e74d0552750/be8552ced8f72971.drcov
lrwxrwxrwx 1 root root   54 2023-10-04 10:07 ./coverage/296faefd3a06d554_09b6e11deb1eaa94.drcov -> ./coverage/296faefd3a06d554/09b6e11deb1eaa94.drcov
lrwxrwxrwx 1 root root   54 2023-10-04 10:07 ./coverage/3b6eda29042c1853_2dbe975e03cb3557.drcov -> ./coverage/3b6eda29042c1853/2dbe975e03cb3557.drcov
lrwxrwxrwx 1 root root   54 2023-10-04 10:07 ./coverage/3ce1da16cb85ea23_862c181171f25614.drcov -> ./coverage/3ce1da16cb85ea23/862c181171f25614.drcov
lrwxrwxrwx 1 root root   54 2023-10-04 10:07 ./coverage/3f13f5ec210f25b5_54a13cf8c0562a51.drcov -> ./coverage/3f13f5ec210f25b5/54a13cf8c0562a51.drcov
lrwxrwxrwx 1 root root   54 2023-10-04 10:07 ./coverage/40d8edcccbd2e06c_d42fe4f80fed8d55.drcov -> ./coverage/40d8edcccbd2e06c/d42fe4f80fed8d55.drcov
lrwxrwxrwx 1 root root   54 2023-10-04 10:07 ./coverage/445c8211b92575cd_6e2273e33924d7ec.drcov -> ./coverage/445c8211b92575cd/6e2273e33924d7ec.drcov
lrwxrwxrwx 1 root root   54 2023-10-04 10:07 ./coverage/4e6e1df4783c9d7d_02fd836017554f62.drcov -> ./coverage/4e6e1df4783c9d7d/02fd836017554f62.drcov
lrwxrwxrwx 1 root root   54 2023-10-04 10:07 ./coverage/52033c22d5db07ca_d75b570e30e7160f.drcov -> ./coverage/52033c22d5db07ca/d75b570e30e7160f.drcov
lrwxrwxrwx 1 root root   54 2023-10-04 10:07 ./coverage/5205d654d2fa198a_f0648c55de769445.drcov -> ./coverage/5205d654d2fa198a/f0648c55de769445.drcov
lrwxrwxrwx 1 root root   54 2023-10-04 10:07 ./coverage/544b30ec1876d6bc_dd250e1cc73068a7.drcov -> ./coverage/544b30ec1876d6bc/dd250e1cc73068a7.drcov
lrwxrwxrwx 1 root root   54 2023-10-04 10:07 ./coverage/63db17375ba3cd79_56c5c7f793cb0364.drcov -> ./coverage/63db17375ba3cd79/56c5c7f793cb0364.drcov
lrwxrwxrwx 1 root root   54 2023-10-04 10:07 ./coverage/68d7fa31c6978f35_d4f79f1f9ad2c86a.drcov -> ./coverage/68d7fa31c6978f35/d4f79f1f9ad2c86a.drcov
lrwxrwxrwx 1 root root   54 2023-10-04 10:07 ./coverage/70dbdd383c271c08_8b56269b1a9a354b.drcov -> ./coverage/70dbdd383c271c08/8b56269b1a9a354b.drcov
lrwxrwxrwx 1 root root   54 2023-10-04 10:07 ./coverage/7d62c4e9d04ed870_2877b333ee32cf73.drcov -> ./coverage/7d62c4e9d04ed870/2877b333ee32cf73.drcov
lrwxrwxrwx 1 root root   54 2023-10-04 10:07 ./coverage/7dbaa59dce446798_b48e177495942de1.drcov -> ./coverage/7dbaa59dce446798/b48e177495942de1.drcov
lrwxrwxrwx 1 root root   54 2023-10-04 10:07 ./coverage/a1462ddb81b6991f_8d97ad9133c8f376.drcov -> ./coverage/a1462ddb81b6991f/8d97ad9133c8f376.drcov
lrwxrwxrwx 1 root root   54 2023-10-04 10:07 ./coverage/a3aecf66a1ab7629_806521f2cb59ccc4.drcov -> ./coverage/a3aecf66a1ab7629/806521f2cb59ccc4.drcov
lrwxrwxrwx 1 root root   54 2023-10-04 10:07 ./coverage/b9cb0d16b96b9605_b811aa4c9eacea97.drcov -> ./coverage/b9cb0d16b96b9605/b811aa4c9eacea97.drcov
lrwxrwxrwx 1 root root   54 2023-10-04 10:07 ./coverage/c429cffca96abd47_13afee6fc790969b.drcov -> ./coverage/c429cffca96abd47/13afee6fc790969b.drcov
lrwxrwxrwx 1 root root   54 2023-10-04 10:07 ./coverage/c565dfc8c9866776_b4756c8f5a5e8fd4.drcov -> ./coverage/c565dfc8c9866776/b4756c8f5a5e8fd4.drcov
lrwxrwxrwx 1 root root   54 2023-10-04 10:07 ./coverage/dabccd1a3b0ad959_7f49497f7f6aee73.drcov -> ./coverage/dabccd1a3b0ad959/7f49497f7f6aee73.drcov
lrwxrwxrwx 1 root root   54 2023-10-04 10:07 ./coverage/ebde2022ad1a59a0_66e6b1baa1df1d90.drcov -> ./coverage/ebde2022ad1a59a0/66e6b1baa1df1d90.drcov
lrwxrwxrwx 1 root root   54 2023-10-04 10:07 ./coverage/f3a25cf58cd8ff28_92b186b4ed4b49b4.drcov -> ./coverage/f3a25cf58cd8ff28/92b186b4ed4b49b4.drcov
lrwxrwxrwx 1 root root   54 2023-10-04 10:07 ./coverage/f9f3781f29afdfc2_d065b483586ca7e7.drcov -> ./coverage/f9f3781f29afdfc2/d065b483586ca7e7.drcov

./coverage/014446d5e22ccab7:
total 8
-rw-rw-rw- 1 root root 375 2023-10-04 10:07 6449d035261ae158.drcov

./coverage/02bd0d6773730397:
total 8
-rw-rw-rw- 1 root root 399 2023-10-04 10:07 990b7a01d108da6e.drcov

./coverage/067ccf3386f6dffd:
total 8
-rw-rw-rw- 1 root root 407 2023-10-04 10:07 fff39e5f01f8ebb1.drcov

./coverage/132e025befe53cb9:
total 8
-rw-rw-rw- 1 root root 375 2023-10-04 10:07 8d7aca8322c310bf.drcov

./coverage/1690bb2862eb85a0:
total 8
-rw-rw-rw- 1 root root 367 2023-10-04 10:07 635d5a3dfd59fcb1.drcov

./coverage/17a47e74d0552750:
total 8
-rw-rw-rw- 1 root root 375 2023-10-04 10:07 be8552ced8f72971.drcov

./coverage/296faefd3a06d554:
total 8
-rw-rw-rw- 1 root root 399 2023-10-04 10:07 09b6e11deb1eaa94.drcov

./coverage/3b6eda29042c1853:
total 8
-rw-rw-rw- 1 root root 367 2023-10-04 10:07 2dbe975e03cb3557.drcov

./coverage/3ce1da16cb85ea23:
total 8
-rw-rw-rw- 1 root root 359 2023-10-04 10:07 862c181171f25614.drcov

./coverage/3f13f5ec210f25b5:
total 8
-rw-rw-rw- 1 root root 359 2023-10-04 10:07 54a13cf8c0562a51.drcov

./coverage/40d8edcccbd2e06c:
total 8
-rw-rw-rw- 1 root root 432 2023-10-04 10:07 d42fe4f80fed8d55.drcov

./coverage/445c8211b92575cd:
total 8
-rw-rw-rw- 1 root root 1937 2023-10-04 10:07 6e2273e33924d7ec.drcov

./coverage/4e6e1df4783c9d7d:
total 8
-rw-rw-rw- 1 root root 480 2023-10-04 10:07 02fd836017554f62.drcov

./coverage/52033c22d5db07ca:
total 8
-rw-rw-rw- 1 root root 1265 2023-10-04 10:07 d75b570e30e7160f.drcov

./coverage/5205d654d2fa198a:
total 8
-rw-rw-rw- 1 root root 367 2023-10-04 10:07 f0648c55de769445.drcov

./coverage/544b30ec1876d6bc:
total 8
-rw-rw-rw- 1 root root 752 2023-10-04 10:07 dd250e1cc73068a7.drcov

./coverage/63db17375ba3cd79:
total 8
-rw-rw-rw- 1 root root 367 2023-10-04 10:07 56c5c7f793cb0364.drcov

./coverage/68d7fa31c6978f35:
total 8
-rw-rw-rw- 1 root root 391 2023-10-04 10:07 d4f79f1f9ad2c86a.drcov

./coverage/70dbdd383c271c08:
total 8
-rw-rw-rw- 1 root root 600 2023-10-04 10:07 8b56269b1a9a354b.drcov

./coverage/7d62c4e9d04ed870:
total 8
-rw-rw-rw- 1 root root 367 2023-10-04 10:07 2877b333ee32cf73.drcov

./coverage/7dbaa59dce446798:
total 8
-rw-rw-rw- 1 root root 375 2023-10-04 10:07 b48e177495942de1.drcov

./coverage/a1462ddb81b6991f:
total 8
-rw-rw-rw- 1 root root 367 2023-10-04 10:07 8d97ad9133c8f376.drcov

./coverage/a3aecf66a1ab7629:
total 8
-rw-rw-rw- 1 root root 375 2023-10-04 10:07 806521f2cb59ccc4.drcov

./coverage/b9cb0d16b96b9605:
total 8
-rw-rw-rw- 1 root root 359 2023-10-04 10:07 b811aa4c9eacea97.drcov

./coverage/c429cffca96abd47:
total 8
-rw-rw-rw- 1 root root 367 2023-10-04 10:07 13afee6fc790969b.drcov

./coverage/c565dfc8c9866776:
total 8
-rw-rw-rw- 1 root root 375 2023-10-04 10:07 b4756c8f5a5e8fd4.drcov

./coverage/dabccd1a3b0ad959:
total 8
-rw-rw-rw- 1 root root 496 2023-10-04 10:07 7f49497f7f6aee73.drcov

./coverage/ebde2022ad1a59a0:
total 8
-rw-rw-rw- 1 root root 375 2023-10-04 10:07 66e6b1baa1df1d90.drcov

./coverage/f3a25cf58cd8ff28:
total 8
-rw-rw-rw- 1 root root 383 2023-10-04 10:07 92b186b4ed4b49b4.drcov

./coverage/f9f3781f29afdfc2:
total 8
-rw-rw-rw- 1 root root 856 2023-10-04 10:07 d065b483586ca7e7.drcov

```
</details>

> It's easy to see which different inputs point to the same coverage

Unfortunately, we can't do that if we use only the name `{coverage_hash}.drcov` for the symlink.

Most of the time, the directory will contain only one file (ideally, it should always be like this), so I think that splitting into directories is already redundant and noisy. In addition, when copying to another machine (for example, from an Android device), all links stop being working and lose their meaning.The version I suggested is essentially the same, but without symbolic links and directories.

What do you think about this, is it still necessary to split into directories?

Reply:
> Unfortunately, we can't do that if we use only the name {coverage_hash}.drcov for the symlink.
why?

Linking `input_hash/coverage_hash.drcov -> ../coverage_hash.drcov` would have the following advantages:
* inputs with multiple coverages -> clear which inputs have low stability, indicating that these have hidden state.
* multiple inputs pointing to the same coverage -> we don't keep duplicated coverage around.

The ideal case of a 1:1 mapping between coverage and inputs is not always the case.

Reply:
> why?

> multiple inputs pointing to the same coverage -> we don't keep duplicated coverage around.

Multiple inputs can point to the same coverage only in one case --- when the coverage is zero. This happens for the reason [that](https://github.com/AFLplusplus/LibAFL/pull/1581#issue-1919797411):

> the coverage is recorded only once --- when instrumenting by Frida (inside [Transformer](https://github.com/AFLplusplus/LibAFL/blob/5b0e3dd3bcbefc4f02e44a6ee0237998cea294ec/libafl_frida/src/helper.rs#L519-L522))

Code instrumentation cannot occur twice, it turns out that the coverage will always be unique, even for the same input running twice (in the second case there will be an empty coverage). <sup>*I hope that I am not mistaken in my judgments*</sup>

> inputs with multiple coverages -> clear which inputs have low stability, indicating that these have hidden state.

Probably, this can be seen quite well with the current implementation without creating directories for each input:

```text
input_hash_0_coverage_hash_0.drcov
input_hash_0_coverage_hash_1.drcov
input_hash_0_coverage_hash_2.drcov
etc.
```

> That being said, this is just an idea, and shouldn't keep this PR back.

If you don't mind, then I would put further discussions in a separate issue/PR, and this PR (if there are no other comments) could be merged.